### ISSUE TITLE
Retrieve parsed portions separately

### ISSUE THREAD

**Issue by vfong3** [Bernardo-MG/dice-notation-java#25]
Functionality of retrieving the number of die, number of sides, operators used, and integers used in operations within the parsed variable would be extraordinarily useful to me. If there is some method in place of retrieving those variables please feel free to show how, as I'd like to use your parser in a discord bot that I am creating (with due credit, of course!)

**Comment by Bernardo-MG** [comment]
Currently this can be done by checking types and casting.

I'll look for a better way to get only a set of dice.

**Comment by vfong3** [comment]
@Bernardo-MG, Thank you for the quick response! If this is indeed a thing that can be done as is I'll certainly try again to implement it, If I somehow get it working through the parser and parsed variables I'll leave more feedback. Good day.

**Comment by vfong3** [comment]
@Bernardo-MG From how it looks it does indeed seem that this should work well with the current system I have in place! I'll attempt implementing it and get back to you as soon as I am able!

Out of curiosity, what library are you using for your random number generation? the typical math library or some other means?

Thank you so much for undertaking this task on such short notice!

**Comment by Bernardo-MG** [comment]
In your example, do you want to keep the "+12" into the result? Meaning, you want it to be added to the result of rolling the dice?

Random numbers are generated by an instance of NumberGenerator. The only implementation of it is RandomNumberGenerator, which makes use of Java's Random.

To change this you would need a custon number generation and something like this:

```java
NumberGenerator numGen;
Roller roller;
DiceNotationExpressionParser parser;

numGen = new CustomNumberGenerator();
roller = new DefaultRoller(numGen);
parser = new DefaultDiceNotationExpressionParser(roller);
```

And what do you mean by the typical format for parsers? If you have a reference documentation I would be glad to take a look and follow any standard which would make the project easier to use.

**Comment by vfong3** [comment]
As it is, I currently use a regex code to return the number of dice and number of sides on those dice, then use those parameters to use a custom random number generator (until I can get a separate RNG library to play along properly) then concatenate that randomly generated number to the old expression

This is the snippet of code that performs that feat:

```
static void run(Message msg)
    {
        FastRand        roller    = new FastRand();  // Custom PRNG
        DoubleEvaluator evaluator = new DoubleEvaluator();  // evaluates infix expression from a string

        String message = msg.getContentRaw().replaceFirst("r ", "").replaceAll(" ", "");
//parses the dice, returning an array with integer values corresponding to the number of die (parsedString[0]) and the number of sides in that die (parsedString[1]).
        int[] parsedString = Dice_Parser.parseDice(message);    

        String rawMessage = message.replaceFirst(String.valueOf(parsedString[0]), "").replaceFirst(
                "d" + String.valueOf(parsedString[1]), ""); // "1d20 + 5"  -> " + 5"
        for (int i = 0; i < parsedString[0]; i++)   //for each die...
        {
            int rollResult = 1 + roller.nextInt(parsedString[1]);   //...roll the die

            String finalExpression = String.valueOf(rollResult) + rawMessage;   //concatenate the result with the previous string
            Double finalResult     = evaluator.evaluate(finalExpression);   //evaluate the full expression

            if (i == 0)
            {
                msg.getChannel().sendMessage(
                        msg.getAuthor().getName() + " rolled a " + message + ": " + rollResult + " + " +
                        parsedString[3] + " = " + finalResult.intValue()).queue(m ->
                                                                          {
                                                                              Timer timer = new Timer();
                                                                              timer.schedule(new TimerTask()
                                                                              {
                                                                                  @Override
                                                                                  public void run()
                                                                                  {
                                                                                      msg.delete().queue();
                                                                                  }
                                                                              }, 0);
                                                                          });
            } else
            {
                msg.getChannel().sendMessage(
                        msg.getAuthor().getName() + " rolled a " + message + ": " + rollResult + " + " +
                        parsedString[3] + " = " + finalResult.intValue()).queue();

            }
        }
    }
```


I would like to replace the parser I am using with your class, so rather than use the array returned i would use your function to get the number of die and the number of sides to use for my custom RNG.

Also, if your full expression evaluator eventually works for all operations (+,-,*,/) then I would use that as well instead of the evaluator I am currently using.

If you have any questions feel free to ask, thank you!

**Comment by Bernardo-MG** [comment]
Sorry I'm asking so many questions, but I want to make sure I give the best solution I can.

I've prepared a parser which returns a single dice set, but the more I think about it the less I'm convinced that will work. What I need to offer is a way to process the output from the parser, not change it.

I'll try to prepare something this week, or next (when I will have more time) at worst.

I'm thinking about something similar to:

```
Iterable<Dice> dice = parsed.getDice();
```

And probably separating the returned object from the parsed tree. And adding a generic way to read it:

```
public <V> V interpret(Interpreter<V> interpreter);
```

In the end, getting the result or getting all the dice (or a single one) is just a way to interpret the result.

**Comment by Bernardo-MG** [comment]
I've just released version 1.3.0.

Now you can acquire the dice from a expression like this:

```java
final TransformableDiceNotationExpression parsed;
final Dice dice;

parsed = new DefaultDiceNotationExpressionParser().parse("1d6");

dice = parsed.transform(new DiceSetsTransformer()).iterator().next();

System.out.println(dice.getQuantity());
System.out.println(dice.getSides());
```
